### PR TITLE
Add the function moving code.

### PR DESCRIPTION
**Pull request by alex4o**
Moving c functions outside their original files, where each c function gets its own file. This allows every function to be in a separate so file. The purpose is to measure the performance degradation of every function being in its own library.

### CODE REVIEW

File: extract.py [hunk 1]
```diff
@@ -4,34 +4,114 @@
...
+            return decl
+        else:
+            return None
+    elif variable.type.kind == TypeKind.TYPEDEF:
+        decl = variable.type.get_declaration()
+        if decl.extent.start.line == 0:
+            print(variable.displayname, variable.extent.start.line)
+        return decl
+    elif variable.type.kind == TypeKind.RECORD:
+        raise NameError("Error: " + str(variable.extent.start.line) + " " + variable )
```
Comment (ltratt) [inline extract.py]: I think `NameError` is meant to mean "you referenced a variable `x` which hasn't been assigned to yet" in Python code itself. I think this is a user-level exception so should be `raise Exception`?

File: extract.py [hunk 2]
```diff
@@ -41,17 +121,30 @@ def get_calls(child):
...
     for file in os.listdir(path):
         if file.endswith(".c"):
-            yield index.parse(path + "/" + file)
+            split_line = 0
+            # this works on the assumption that all includes are at the start of the file.
+            # calculating the split line gives information about which things are defined in
+            # headers and which are defined in in the c file.
+            with open(path + "/" + file, "r") as f:
+                for current_line_no, line in enumerate(f):
+                    match = re.match("# [0-9]+ \"" + file + "\" 2", line)
```
Comment (ltratt) [inline extract.py:130]: Is the `[0-9]+` to work with generated `#`s from the preprocessor?

File: main.py [hunk 3]
```diff
@@ -5,13 +5,80 @@
...
+    if args.graph:
+        graph = extract.generate_graph(args.path)
+
+        bc = graph.biconnected_components()
+        igraph.plot(graph, target="bc.svg", bbox=(0, 0, 9200, 9200), vertex_label=graph.vs["short"], mark_groups=bc)
+
+        ceb = graph.community_edge_betweenness(directed=False)
+        igraph.plot(graph, target="ceb.svg", bbox=(0, 0, 9200, 9200), vertex_label=graph.vs["short"], mark_groups=ceb.as_clustering())
+    else:
+        """
```
Comment (ltratt) [inline main.py]: Probably should be a `#` comment rather than a sort-of doc string?

File: main.py [hunk 4]
```diff
@@ -5,13 +5,80 @@
...
+            file = open(tu.spelling, "r+")
+            for (k, v) in extract.get_decl(tu, split_line, CursorKind.VAR_DECL):
+                calls = ""
+
+                for call in extract.get_calls(v, CursorKind.DECL_REF_EXPR):
+                    calls += "extern " + call.referenced.result_type.spelling + " " + call.referenced.displayname + ";\n"
+
+                var_text = extract.file_read_extent(file, v.extent)
+
+                res_file = open(args.output + "/" + v.spelling + ".c", "w")
```
Comment (ltratt) [inline main.py]: `os.path.join` is a nicer way of concatenating dir / path names.

File: extract.py (hunk 1 above)
Comment (alex4o) [reply extract.py]: I agree, I wanted to throw a different exception but the generic one should be sufficient.

File: extract.py (hunk 2 above)
Comment (alex4o) [reply extract.py:130]: Yes
```c
# 34 "ldo.c" 2
# 84 "ldo.c"
```
This is how they look like, and can be used to identify the different portions of code that comprise the final source file.

Comment (ltratt) [reply extract.py:130]: Got it!

File: main.py (hunk 3 above)
Comment (alex4o) [reply main.py]: From my understanding """ are the only way to create mulltiline comments and the only official one, but that shouldn't be a problem.

Comment (ltratt) [reply main.py]: `"""` in this sense is for doc comments for methods (etc).: for "normal" comments you just have to use `#` repeated for each line, I'm afraid.

Comment (alex4o) [reply main.py]: Yes this is how I did it in the end.

https://github.com/capablevms/function-community-finding/blob/c597ed8134fe6b07a4a40eec7d5606d508704037/main.py#L21-L23

**Comment by alex4o** [comment]
Should be ready for merging.

File: extract.py [hunk 5]
```diff
@@ -125,12 +125,12 @@ def get_tus(index, path):
             # this works on the assumption that all includes are at the start of the file.
             # calculating the split line gives information about which things are defined in
             # headers and which are defined in in the c file.
-            with open(path + "/" + file, "r") as f:
+            with open(os.path.join(path ,file), "r") as f:
```
Comment (ltratt) [inline extract.py]: `(path, file)` (space *after* the comma) conventionally?

Comment (alex4o) [reply extract.py]: Thank you for noticing that. Fixed https://github.com/capablevms/function-community-finding/pull/2/commits/7d96f57a92a808e32f5a849e773177a3636e36b1.

**Comment by ltratt** [comment]
Please squash.

**Comment by alex4o** [comment]
Done

**Comment by ltratt** [comment]
Minor commit typo:

>The function moving code attempts to move each function into it's own

No `'` in "it's" here. Please feel free to force push an update.

**Comment by alex4o** [comment]
Fixed

**Pull request merged**